Fix renegotiation with Client Auth

mattcaswell opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->

#1920 describes a problem where a non-client auth reneg handshake that occurs after a client-auth handshake will end up with the client sending a Certificate message when it is not supposed to.

There is another related bug where, in the same situation, the server expects a Certificate message even when it hasn't requested one!

These two bugs together mean that OpenSSL will happily talk to the same version of itself in this scenario, but nothing else.

In adding tests for this I found a third bug in the implementation of SSL_VERIFY_CLIENT_ONCE. The documentation for this flag says:
````
      B<Server mode:> only request a client certificate on the initial TLS/SSL
      handshake. Do not ask for a client certificate again in case of a
      renegotiation. This flag must be used together with SSL_VERIFY_PEER.
    
      B<Client mode:> ignored
````
But the implementation actually did nothing. After the server sends its
ServerKeyExchange message, the code was checking s->session->peer to see if
it is NULL. If it was set then it did not ask for another client
certificate. However s->session->peer will only be set in the event of a
resumption, but a ServerKeyExchange message is only sent in the event of a
full handshake (i.e. no resumption).
    
The documentation suggests that the original intention was for this to
have an effect on renegotiation, and resumption doesn't come into it.
    
The fix is to properly check for renegotiation, not whether there is already
a client certificate in the session.
    
As far as I can tell this has been broken for a *long* time. I actually wondered whether we would be better off just removing it. But as it turns out I need to use it for my tests to work so I fixed it anyway!!

I also fixed a bug in TLSProxy where zero length messages weren't being recorded properly.

This is for master. I will create a separate PR for 1.1.0

Fixes #1920

kroeckx commented:
Can this change break something for old clients or servers talking to a fixed client or server?

mattcaswell commented:
If an old 1.1.0 OpenSSL client talks to a fixed 1.1.0 OpenSSL server and it does a ClientAuth handshake followed by a non-client auth handshake then it will fail (and vice versa with a fixed client and an old server). Whereas if an old 1.1.0 client talks to an old 1.1.0 server then it will work. I don't think that issue is worth worrying about.

mattcaswell commented:
I rebased this as it was showing a conflict with current master. Having rebased I found that one of the TLSv1.3 tests were failing. This is due to the way this MR was detecting a reneg (i.e. checking whether the encryption/hash context are non NULL). In TLSv1.3 this happens a lot earlier. I changed the test to use finish_md_len (which we already use elsewhere to test for reneg, so its probably better anyway).

Ping?

levitte commented:
This could do with a rebase on a fresher master...  or get closed if it has expired somehow.

mattcaswell commented:
@levitte. I rebased this as requested. Please also look at the related PRs #1983 and #1984.

levitte commented on ssl/statem/statem.c:
```diff
@@ -347,14 +347,16 @@ static int state_machine(SSL *s, int server)
             s->hit = 0;
 
-            s->s3->tmp.cert_request = 0;
+            s->s3->tmp.cert_req = 0;
```
Side note: we should consider renaming `cert_req` or `cert_request`.  I cannot easily see how one differs from the other except by name.

levitte commented:
Looks good to me.  Let's see the build results.

mattcaswell commented:
@levitte - build results were good.

mattcaswell commented:
@levitte - any chance you could look at #1983 and #1984, as well which are closely related?

levitte commented:
Done

mattcaswell commented:
Pushed. Thanks.
